File: lib/translate.js
const PUBLIC_INSTANCES = [
  'https://libretranslate.com',
  'https://translate.argosopentech.com',
  'https://translate.fedilab.app',
];

let currentInstance = 0;

async function translateText(text, source = 'ar', target = 'en') {
  if (!text || text.trim().length === 0) return text;

  // Skip if mostly ASCII (already English)
  const asciiCount = (text.match(/[\x00-\x7F]/g) || []).length;
  if (asciiCount / text.length > 0.7) return text;

  for (let attempt = 0; attempt < PUBLIC_INSTANCES.length; attempt++) {
    const idx = (currentInstance + attempt) % PUBLIC_INSTANCES.length;
    const instance = PUBLIC_INSTANCES[idx];

    try {
      const res = await fetch(`${instance}/translate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          q: text,
          source,
          target,
          format: 'text',
        }),
      });

      if (!res.ok) continue;

      const data = await res.json();
      if (data && data.translatedText) {
        currentInstance = idx;
        return data.translatedText;
      }
    } catch {
      continue;
    }
  }

  return text;
}

async function translateTitle(parsed) {
  // If regex already extracted English show name, return it
  if (parsed.english && !/[\u0600-\u06FF]/.test(parsed.english)) {
    return parsed;
  }

  // Try translating the full clean title
  const translated = await translateText(parsed.clean);
  if (translated !== parsed.clean) {
    return { ...parsed, english: translated };
  }

  return parsed;
}

export { translateText, translateTitle };

File: egydead/translate.py
import json
import re
import urllib.request
import urllib.error

PUBLIC_INSTANCES = [
    'https://libretranslate.com',
    'https://translate.argosopentech.com',
    'https://translate.fedilab.app',
]

_current_instance = 0


async def translate_text(text: str, source: str = 'ar', target: str = 'en') -> str:
    if not text or not text.strip():
        return text

    ascii_count = len(re.findall(r'[\x00-\x7F]', text))
    if ascii_count / max(len(text), 1) > 0.7:
        return text

    global _current_instance
    for attempt in range(len(PUBLIC_INSTANCES)):
        idx = (_current_instance + attempt) % len(PUBLIC_INSTANCES)
        instance = PUBLIC_INSTANCES[idx]

        try:
            body = json.dumps({
                'q': text,
                'source': source,
                'target': target,
                'format': 'text',
            }).encode()

            req = urllib.request.Request(
                f'{instance}/translate',
                data=body,
                headers={'Content-Type': 'application/json'},
                method='POST',
            )
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = json.loads(resp.read().decode())
                if data and data.get('translatedText'):
                    _current_instance = idx
                    return data['translatedText']
        except (urllib.error.URLError, json.JSONDecodeError, OSError):
            continue

    return text


async def translate_title(parsed: dict) -> dict:
    english = parsed.get('english', '')
    if english and not re.search(r'[\u0600-\u06FF]', english):
        return parsed
    translated = await translate_text(parsed.get('clean', ''))
    if translated != parsed.get('clean'):
        return {**parsed, 'english': translated}
    return parsed
